gobblin-core/src/main/java/org/apache/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -352,4 +377,58 @@ public void handleMessage(ControlMessage message) {
       cloner.close();
     }
   }
+
+  /**
+   * Get the serialized key to partitions info in {@link #state}
+   */
+  private static String getPartitionsKey(int branchId) {
+    return String.format("writer.%d.partitions", branchId);
+  }
+
+  /**
+   * Serialize partitions info to {@link #state} if they are any
+   */
+  private void maySerializePartitions() {
+    List<PartitionDescriptor> descriptors = new ArrayList<>();
+
+    for (DataWriter writer : partitionWriters.asMap().values()) {
+      Descriptor descriptor = writer.getDataDescriptor();
+      if (null == descriptor) {
+        log.warn("Drop partition info as writer {} returns a null PartitionDescriptor", writer.toString());
+        continue;
+      }
+
+      if (!(descriptor instanceof PartitionDescriptor)) {
+        log.warn("Drop partition info as writer {} does not return a PartitionDescriptor", writer.toString());
+        continue;
+      }
+
+      descriptors.add((PartitionDescriptor)descriptor);
+    }
+
+    if (descriptors.size() > 0) {
+      state.setProp(getPartitionsKey(branchId), GSON.toJson(descriptors, PARTITION_LIST_TYPE));
+    } else {
+      log.info("Partitions info not available. Will not serialize partitions");
+    }
+  }
+
+  /**
+   * Get the partitions info of a work unit from the {@code state}. Then partitions info will be removed from the
+   * {@code state}. A
+   *
+   * <p>
+   *   In Gobblin, only the {@link PartitionedDataWriter} knows all partitions written for a work unit. Each partition
+   *   {@link DataWriter} decides the actual form of a dataset partition
+   * </p>
+   */
+  public static List<PartitionDescriptor> getPartitionsAndClean(State state, int branchId) {
+    String partitionsKey = getPartitionsKey(branchId);
+    String json = state.getProp(partitionsKey);
+    if (Strings.isNullOrEmpty(json)) {
+      return Lists.newArrayList();
+    }
+    state.removeProp(partitionsKey);
+    return GSON.fromJson(json, PARTITION_LIST_TYPE);
```

Comment:
It will throw RuntimeExceptions. I expect the job to fail in this case so that we won't report wrong lineage and it raises an explicit signal of a bug to fix.

Reply:
ok, works for me